ppa/archive/management/commands/hathi_images.py

```diff
@@ -0,0 +1,286 @@
+        return report
+
+
+class Command(BaseCommand):
+    """
+    Download HathiTrust page image data via image server
+
+    Note: Excerpts cannot be specified individually, only by source (collectively)
+    """
+    help = __doc__
+
+    # Interrupt flag to exit gracefully (i.e. between volumes) when a signal is caught
+    interrupted = False
+
+    # Argument parsing
+    def add_arguments(self, parser):
+        """
+        Configure additional CLI arguments
+        """
+        parser.add_argument(
+            "output_dir",
+            type=Path,
+            help="Top-level output directory"
+        )
+        parser.add_argument(
+            "--htids",
+            nargs="+",
+            help="Optional list of HathiTrust ids (by default, downloads images for all public HathiTrust volumes)",
+        )
+        parser.add_argument(
+            "--image-width",
+            type=int,
+            help="Width for full-size images in pixels. Default: 800",
+            default=800,
+        )
+        parser.add_argument(
+            "--thumbnail-width",
+            type=int,
+            help="Width for thumbnail images in pixels. Must be at most 250 pixels. Default: 250",
+            default=250,
+        )
+        parser.add_argument(
+            "--progress",
+            action=argparse.BooleanOptionalAction,
+            help="Display progress bars to track download progress",
+            default=True,
+        )
+  
+    def interrupt_handler(self, signum, frame):
+        """
+        For handling of SIGINT, as possible. For the first SIGINT, a flag is set
+        so that the command will exit after the current volume's image download
+        is complete. Additionally, the default signal handler is restored so a
+        second SIGINT will cause the command to immediately exit.
+        """
+        if signum == signal.SIGINT:
+            # Restore default signal handler
+            signal.signal(signal.SIGINT, signal.SIG_DFL)
+            # Set interrupt flag
+            self.interrupted = True
+            self.stdout.write(self.style.WARNING(
+                "Command will exit once this volume's image download is "
+                "complete.\nCtrl-C / Interrupt to quit immediately"
+                )
+            )
+
+    def download_image(self, page_url: str, out_file: Path) -> bool:
+        """
+        Attempts to download and save an image from the specified URL.
+        Returns a boolean corresponding to whether the download was successful
+        """
+        response = requests.get(page_url)
+        success = False
+        if response.status_code == requests.codes.ok:
+            with out_file.open(mode="wb") as writer:
+                writer.write(response.content)
+            success = True
+        elif response.status_code == 503:
+            logger.debug("Received 503 status code. Throttling may have occurred")
+        return success
```

Comment:
Nice. Have you run into any other error codes in your testing? It would be good to log other errors, but if they haven't come up yet then let's leave it.

Reply:
I haven't actually run into any of these errors.